PERFORMANCE.md

```diff
@@ -0,0 +1,80 @@
+# Performance
+
+It is important to be conscious of performance in hot code paths. In such code paths there should be 3 goals:
+
+- Be as fast as possible
+- Be memory efficient
+- Be as stable as possible
+
+This guide is meant to document known performance issues so that the contributors are aware of them.
+
+## Performance Sensitive Areas
+
+- Change detection: `observer`, `computed`
+- View creation: `View`, `Renderer`
+- Routing: `Router`
+- Dependency Injection: `Container`, `Registry`
+
+__Note :__ If you are not in a performance sensitive area feel free to use these constructs. That being said, please do your due diligence to ensure your contribution is indeed not in a hot path.
+
+## Array Methods
+
+Most of the time using array methods e.g. `forEach`, `map`, `filter`, is fine, but in hot paths these methods can become very costly. One would think that these methods would almost be like macros that would compile into more efficient code, but that is not the case. In hot paths these methods can be up to 10X slower than just writing a `for`-loop.
+
+One of the reasons behind the slowness is largely do to the fact that these methods take a closure which allocates memory. In turn this creates GC pressure and potentially creates a polymorphic call site. For a more in depth explanation please see the following [explainer](https://gist.github.com/stefanpenner/376efb906388954146a8).
+
+## Closures
+
+Try to keep closures to a minimum in hot code paths.
+
+- They create GC pressure.
+- They may result in polymorphic call site (see below).
+
+The reason behind why closures are slower is that contrary to popular belief, closures actually cause the VM to create a stateful `context` object at the call site. This is very similar to `new`-ing a constructor and creating an instance at the call site. Because of this closures are grouped into the same Polymorphic and Monomorphic call site heuristics.
+
+## Polymorphic vs Monomorphic Call Sites
+
+Calling `fn()` may be fast or slow depending whether the `fn` variable is polymorphic (changes over time) or monomorphic (stable). The difference can be 4x in speed. Monomorphic sites can be inlined, where as polymorphic sites prevent inline and may even cause de-optimization of your code. In general stick to classes with instance method, and avoid polymorphism for code sensitive paths such as change-detection and dependency injection. For more information about how VMs handle optimizations please read [this article](http://mrale.ph/blog/2015/01/11/whats-up-with-monomorphism.html).
+
+```
+function add(x, y) {
+  return x + y;
+}
+
+add(1, 2);      // + in add is monomorphic
+add("a", "b");  // + in add becomes polymorphic
+```
+
+## Object Shaping And Stability
+
+VMs optimize JS objects by creating hidden classes for each property of that object or context. Because of this you should avoid mutating the shape of an object once it is created.
+
+```
+class Point { // Hidden class 0
+  constructor(x) {
```

Comment:
Mention conditional assignment should still result in the same shape per branch.  Also, reading from `this.someProp` that falls through to the prototype then assigning `this.someProp = val` means that `this` requires 2 shapes and makes the method not optimizable, since hidden classes are flattened.  On the flattened note, `in`, `hasOwnProperty` and for in loops on objects with prototype other than Object.prototype are slow because they can't use the flattened version.